B2 efficiency

fracai opened the pull request:
As discussed in #179 these commits use _b2_download_file_by_name_ instead of _b2_list_file_names_ to retrieve chunk information.
The switch only occurs when a single file is being checked without previous versions (fossils). The call to download doesn't return the "**B2 action**", so the code just supplies "**upload**". This matches the current expectation as a hidden chunk would not be visible to _b2_list_file_names_ either. A fossil would be visible to _b2_list_file_versions_, but that is only called if _includeVersions_ is **true**.
I haven't seen any noticeable impact to run time, but I can see the change in recorded transaction classes on the B2 summary.

The only part that feels outright hacky to me is the switch statement at the top of _call_. I made the http call a HEAD request (B2 could at some point might optimize their API to not charge the requested bytes) and thus needed to pass in a different _input_ paramater. I picked _bool_. This does work, but I wonder if it would be better supported by adding another paramater to specify the request type? Or an enum to make that switch statement more explicit?

gilbertchen commented:
Thank you for the work.  I will review the changes tomorrow and may take a couple of days to run some tests.

fracai commented:
Sure, let me know if you have any architecture preferences. After putting this together I had the thought that it might be better served in its own method, but figured that was getting more towards style preferences.
I hope it proves useful!

gilbertchen commented on src/duplicacy_b2client.go:
```diff
@@ -81,7 +81,7 @@ func (client *B2Client) retry(backoff int, response *http.Response) int {
 }
 
-func (client *B2Client) call(url string, input interface{}) (io.ReadCloser, int64, error) {
+func (client *B2Client) call(url string, requestHeaders map[string]string, input interface{}) (io.ReadCloser, http.Header, int64, error) {
```
It may be better to add the method as an argument and not pass requestHeaders.

fracai replied:
Is there any desire to keep requestHeaders as an argument in the event that other calls need to pass anything in the future? Looking at the B2 API it looks like only b2_upload_file and b2_download_file_by_name take a request header other than authorization.

I'll make the method an enum and construct the header in that switch.

gilbertchen commented on src/duplicacy_b2client.go:
```diff
@@ -115,6 +118,12 @@ func (client *B2Client) call(url string, input interface{}) (io.ReadCloser, int6
 			request.Header.Set("Authorization", client.AuthorizationToken)
 		}
 
+		if requestHeaders != nil {
```
The header `Range: bytes=0-0` can be added here if the method is "GET".

gilbertchen commented on src/duplicacy_b2client.go:
```diff
@@ -131,11 +140,11 @@ func (client *B2Client) call(url string, input interface{}) (io.ReadCloser, int6
-			return response.Body, response.ContentLength, nil
+			return response.Body, response.Header, response.ContentLength, nil
 		}
 
```
I think you need to check if the response code is 404 and the method is HEAD.  If so return nil httpHeader.  Without this check it will keep retrying when looking up a non-existent file.

gilbertchen commented on src/duplicacy_b2client.go:
```diff
@@ -299,6 +323,18 @@ func (client *B2Client) ListFileNames(startFileName string, singleFile bool, inc
 
 		output := B2ListFileNamesOutput{}
 
+		if !inputMapBool && singleFile && !includeVersions {
```
Again here you need to check if responseHeader is nil.

fracai replied:
I'll add that and some other error checking to be sure the expected headers are present.

gilbertchen commented:
Here is how I run the tests.  First create a file named `test_storage.conf` under `.\src`:
```
{
    "b2": {
        "bucket": "bucket_name",
        "account": "b2_account",
        "key": "b2_key"
    }
}
```
Then run the following tests:

```
go test -run TestB2Client
go test -run TestStorage ./src -storage b2
go test -run TestBackupManager ./src -storage b2
```

tbain98 commented on src/duplicacy_b2client.go:
```diff
@@ -284,13 +293,28 @@ func (client *B2Client) ListFileNames(startFileName string, singleFile bool, inc
 	input["bucketId"] = client.BucketID
 	input["startFileName"] = startFileName
 	input["maxFileCount"] = maxFileCount
+	inputMapBool := true
```
I think this will disappear if you take my suggestion to pass the HTTP verb in as an argument, but this isn't a very approachable variable name for anyone seeing this code without a pre-existing understanding of what it does (not least because a boolean's name should represent what will happen if it is true, but this variable's name implies that you'll pass a boolean in place of the input map argument, but then does the opposite). I hope you'll be eliminating it entirely (setting method = "HEAD" and input = nil, for example), but if you were going to keep it, a better name would be something like passBoolToInputParameter.

fracai replied:
all valid

tbain98 commented on src/duplicacy_b2client.go:
```diff
@@ -299,6 +323,18 @@ func (client *B2Client) ListFileNames(startFileName string, singleFile bool, inc
 
 		output := B2ListFileNamesOutput{}
 
+		if !inputMapBool && singleFile && !includeVersions {
```
Do we need to test !inputMapBool here? Is there a situation where we'd get here and singleFile would be true, includeVersions would be false, and inputMapBool would be true? Seems like unnecessary complexity to the if statement.

fracai replied:
You're right, it's there as a sanity check, but it is redundant. That bool is going away regardless. I'll see if it makes sense to have anything similar after the other changes.

fracai commented on src/duplicacy_b2client.go:
```diff
@@ -94,19 +94,22 @@ func (client *B2Client) call(url string, input interface{}) (io.ReadCloser, int6
 		case int:
 			method = "GET"
 			inputReader = bytes.NewReader([]byte(""))
+		case bool:
```
You're right, there are the constants in the http package. Those should be used here. https://golang.org/pkg/net/http/#pkg-constants

More generally, I think this can be cleaner by passing in a map of headers, input as a bytes.Reader, and the method as string (a switch can restrict this to only match the http constants)

```
call(url string, requestMethod string, requestHeaders map[string]string, input *bytes.Reader)

…

switch requestMethod {
    default:
        return nil, nil, 0, err
    case MethodGet:
        break
    case MethodHead:
        break
    case MethodPost:
        break
}
```

tbain98 replied:
What you've proposed for the request method sounds fine given that strongly-typed enums don't exist in Go.

I'm not wild about the suggestion that creating the bytes.Reader should be done each place that the method is called; that feels like it'll lead to code duplication. Rather, keep the current parameter type of interface, update the logic to allow passing nil for the cases where empty content should be used for the request, and remove the special cases for int and bool (passing nil instead).

fracai replied:
I started thinking about the duplication as well. One idea was making use of lambdas. I haven't had a chance to actually implement anything yet and lambdas seem heavy handed here as well. I'll try your suggestion.

fracai commented:
I've cleaned up the method specification and added a bunch of error checking.
I'm still passing in the requestHeaders as it really doesn't seem clean to construct those in the _call_ method, but I can if that's desired.
I haven't had a chance to run the formal tests yet, but it does seem fine based on a few manual runs.

gilbertchen commented on src/duplicacy_b2client.go:
```diff
@@ -144,15 +159,24 @@ func (client *B2Client) call(url string, input interface{}) (io.ReadCloser, int6
 			continue
+		} else if response.StatusCode == 404 {
+			if http.MethodHead == method {
+				return nil, nil, 0, fmt.Errorf("URL request '%s' returned status code %d", url, response.StatusCode)
```
Can't return an error here.  Should instead tell the caller that the given file doesn't exist in the storage.

fracai replied:
I'm not sure I follow. If I look at how the code is used, returning an error does indicate that the file doesn't exist.

b2storage.GetFileInfo for example

Are you saying it should just be swallowing the error and returning "nil, nil, 0, nil"?

gilbertchen replied:
Yes, if the file to be looked up doesn't exist, GetFileInfo expects ListFileNames to return an empty Entry array, so maybe here should return a responseHeader of nil and in ListFileNames use that as the condition to return the empty Entry array.

fracai replied:
I've updated the 404 branch to return a nil error and then ListFileNames treats that nil responseHeader as a missing file and returns an empty array.

gilbertchen commented:
Thank you for your contribution!
